source/_posts/2019-07-15-complete-flows-partial-models.markdown

```diff
@@ -0,0 +1,166 @@
+---
+layout: post
+author: rafa
+title: "(Complete, Flows) -> (Partial) -> (Models)"
+date: 2019-07-15 19:55:56 +0200
+comments: true
+categories: 
+---
+
+Most apps these days have a sequence of screens that gather information from the user, like a registration flow, a form of some kind. The data from each step is typically combined into a single data structure.
+For example, let's say we want the name, age, and the password to authenticate the user.
+<!--more-->
+One way to model it is by using the following data structure:
+
+```swift
+struct FormData {
+    let name: String
+    let age: Int
+    let password: String
+}
+```
+
+One issue we are going to come about is that our model is strict, it needs all the values at once, whereas users will supply each value at a time. First they will type in their name, then their age, and so on.
+Wrapping up the fields in `Optional`'s, may loosen it's strictness.
+
+```swift
+struct FormData {
+    let name: String?
+    let age: Int?
+    let password: String?
+}
+```
+
+Our flow code might look like:
+
+```swift
+func firstStepFinished(with name: String) -> FormData {
+    return FormData(name: name, age: nil, password: nil)
+}
+
+func secondStepFinished(with age: Int, partialFormData: FormData) -> FormData {
+    return FormData(name: partialFormData.name, age: age, password: nil)
+}
+
+func thirdStepFinished(with password: String, partialFormData: FormData) {
+    let formData = FormData(name: partialFormData.name, age: partialFormData.age, password: password)
+    
+    api.performLogin(with: formData)
+}
+```
+
+However, now we need to `guard` against any `nil` values if we want to use them (for example, to make a network request).
+
+```swift
+guard let name = formData.name,
+      let age = formData.age,
+      let password = formData.password {
+    return // what should we do here???
+}
+
+// use data
+```
+
+From a domain perspective, that `return` doesn't make any sense.
+One may say, _"we can raise an error to the user"_ or _"we could track it and check if users are getting stuck somehow"_. But, at the end of the day, this is not a good solution because you know that when the flow ends, you have all the values.
+
+Our model is "lying" to us. That's not loosen, it's just flawed.
+
+There are several approaches to make it better, like "one model _per_ step":
+
+```swift
+struct FirstStep {
+    let name: String
+}
+
+struct SecondStep {
+    let name: String
+    let age: Int
+    
+    init(firstStep: FirstStep, age: Int) {
+        self.name = firstStep.name
+        self.age = age
+    }
+}
+
+struct ThirdStep {
+    let name: String
+    let age: Int
+    let password: String
+    
+    init(secondStep: SecondStep, password: String) {
+        self.name = secondStep.name
+        self.age = secondStep.age
+        self.password = password
+    }
+}
+```
+
+That's better! But there is also another way of doing things that that doesn't involve duplication and of partial data structs.
```

Comment:
1. there's a `that that` here
2. doesn't involve duplication `nor` partial data structs